generator/template.go

```diff
@@ -195,6 +195,91 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
     }
   })
 }
+
+/**
+ * Flattens a deeply nested object
+ *
+ * @param  {Object} deeplyNestedObject
+ * @param  {String} path
+ * @return {Record<string, unknown>}
+ */
+function flattenObject<T extends Record<string, unknown>>(
+  deeplyNestedObject: T,
+  path: string = ""
+): T {
+  return Object.keys(deeplyNestedObject).reduce(
+    (acc: T, key: string): T => {
+      const value = deeplyNestedObject[key];
+
+      const newPath = Array.isArray(deeplyNestedObject)
+        ? path + "." + key
+        : [path, key].filter(Boolean).join(".");
+      const isObject = [
+        typeof value === "object",
+        value !== null,
+        !(value instanceof Date),
+        !(value instanceof RegExp),
+        !(Array.isArray(value) && value.length === 0)
+      ].every(Boolean);
+
+      return isObject
+        ? { ...acc, ...flattenObject(value as Record<string, unknown>, newPath) }
+        : { ...acc, [newPath]: value };
+    },
+    {} as T
+  );
+}
+
+/**
+ * Renders a deeply nested object into a string of URL search parameters
+ * by first flattening the object and then removing non-primitive array keys,
+ * non-primitive values and keys which are already present in the URL path.
+ * @param  {Object} deeplyNestedObject
+ * @param  {Array<string>} urlPathParams
+ * @return {String}
+ */
+export function renderURLSearchParams<T extends Record<string, unknown>>(
+  deeplyNestedObject: T,
+  urlPathParams: string[] = []
+): string {
+  const flattenedObject = flattenObject(deeplyNestedObject);
+
+  const urlSearchParams = Object.keys(flattenedObject).reduce(
+    (acc: string[][], key: string): string[][] => {
+      const parts = key.split(".");
+      const index = parts.findIndex(f => Number(f) || f === "0");
+
+      // if key does not contain only one numeric index as the last element
+      // then it is an array of objects
+      if (parts.length === index + 1 || index === -1) {
+        // remove array index to render url search params properly
+        // according to http.proto
+        // https://github.com/googleapis/googleapis/blob/master/google/api/http.proto
+        const keyWithoutArrayIndex =
+          index > -1 ? parts.slice(0, index).join(".") : key;
+
+        // values should only be primitive types and key should not be
+        // present in the url path as a parameter
+        const value = flattenedObject[key];
+        if (
+          (typeof value === "string" ||
+            typeof value === "boolean" ||
+            typeof value === "number") &&
+          !urlPathParams.find(f => f === key)
+        ) {
+          acc = [...acc, [keyWithoutArrayIndex, value.toString()]];
+        }
+      }
+
+      return acc;
+    },
+    [] as string[][]
+  );
+
+  return urlSearchParams.length > 0
```

Comment:
wouldn't mind to have just `new URLSearchParams(urlSearchParams).toString()` here

Reply:
Thank you! I will fix this as per your suggestion 🙇🏼